File: solutions/2244. Minimum Rounds to Complete All Tasks/solution.py
class Solution(object):
    def minimumRounds(self, tasks):
        """
        :type tasks: List[int]
        :rtype: int
        """
        minRounds = 0
        roundCosts = [-1,-1,1,1]
        tasks = sorted(tasks)
        countTasks = {
            tasks[0]: 0,
        }
        for i in range(len(tasks)):
            countTasks[tasks[i]] += 1
            if i < len(tasks)-1 and tasks[i] != tasks[i+1]:
                countTasks[tasks[i+1]] = 0
            if i == len(tasks)-1 or tasks[i] != tasks[i+1]:
                if countTasks[tasks[i]] >= len(roundCosts):
                    for j in range(len(roundCosts),countTasks[tasks[i]]+1):
                        if roundCosts[j-2] < roundCosts[j-3]:
                            roundCosts.append(roundCosts[j-2]+1)
                        elif roundCosts[j-3] > 0:
                            roundCosts.append(roundCosts[j-3]+1)
                        else:
                            roundCosts.append(roundCosts[j-2]+1)
                if roundCosts[countTasks[tasks[i]]] == -1:
                    return -1
                else:
                    minRounds += roundCosts[countTasks[tasks[i]]]
        return minRounds

File: solutions/2244. Minimum Rounds to Complete All Tasks/solution.cpp
class Solution {
public:
    int minimumRounds(vector<int>& tasks) {
        map<int, int> countTasks;
        vector<int> roundCosts{-1,-1,1,1};
        set<int> allTasks;
        for(int i = 0; i < tasks.size(); i++){
            if (countTasks.find(tasks[i]) != countTasks.end()) {
                countTasks[tasks[i]]++;
            }
            else{
                countTasks[tasks[i]] = 1;
            }
            if(!allTasks.contains(tasks[i])){
                allTasks.insert(tasks[i]);
            }
            if (countTasks[tasks[i]]>= roundCosts.size()){
                for(int j = roundCosts.size(); j < countTasks[tasks[i]]+1; j++){
                    if (roundCosts[j-2]<roundCosts[j-3]){
                        roundCosts.push_back(roundCosts[j-2]+1);
                    }
                    else if (roundCosts[j-3] > 0){
                        roundCosts.push_back(roundCosts[j-3]+1);
                    }
                    else {
                        roundCosts.push_back(roundCosts[j-2]+1);
                    }
                }
            }
        }
        int res = 0;
        for(set<int>::iterator it = allTasks.begin(); it != allTasks.end(); it++){
            if(roundCosts[countTasks[*it]] == -1){
                return -1;
            }
            else{
                res += roundCosts[countTasks[*it]];
            }
        }
        return res;
    }
};
